File: app/src/main/java/com/work/lazxy/writeaway/mvpframe/concrete/planning/PlanningModel.java
package com.work.lazxy.writeaway.mvpframe.concrete.planning;


import android.util.SparseArray;

import com.work.lazxy.writeaway.db.PlanningDataHandler;
import com.work.lazxy.writeaway.entity.PlanningEntity;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.annotations.NonNull;

/**
 * Created by Lazxy on 2017/2/22.
 */

public class PlanningModel implements PlanningContract.Model {

    @Override
    public Observable<List<PlanningEntity>> getPlanning() {
        return Observable.create(new ObservableOnSubscribe<List<PlanningEntity>>() {
            @Override
            public void subscribe(@NonNull ObservableEmitter<List<PlanningEntity>> e) throws Exception {
                List<PlanningEntity> plannings = new ArrayList<>();
                SparseArray<String> rawData = PlanningDataHandler.getInstance().getPlannings();
                if(rawData!=null&&rawData.size()>0){
                    for(int i=1;i<=rawData.size();i++){
                        plannings.add(new PlanningEntity(i,rawData.get(i)));
                    }
                }
                e.onNext(plannings);
                e.onComplete();
            }
        });
    }

    @Override
    public Observable<String> savePlanning(final List<PlanningEntity> plannings) {
        return Observable.create(new ObservableOnSubscribe<String>() {
            @Override
            public void subscribe(@NonNull ObservableEmitter<String> e) throws Exception {
                PlanningDataHandler.getInstance().savePlannings(plannings);
                e.onNext("");
                e.onComplete();
            }
        });
    }

    @Override
    public Observable<String> addPlanning(final PlanningEntity planning) {
        return Observable.create(new ObservableOnSubscribe<String>() {
            @Override
            public void subscribe(@NonNull ObservableEmitter<String> e) throws Exception {
                PlanningDataHandler.getInstance().addPlanning(planning);
                e.onNext("");
                e.onComplete();
            }
        });
    }
}

File: app/src/main/java/com/work/lazxy/writeaway/mvpframe/concrete/planning/PlanningModel.kt
package com.work.lazxy.writeaway.mvpframe.concrete.planning

import com.work.lazxy.writeaway.db.PlanningDataHandler
import com.work.lazxy.writeaway.entity.PlanningEntity
import io.reactivex.Observable
import java.util.*

/**
 * Created by Lazxy on 2017/2/22.
 */
class PlanningModel : PlanningContract.Model {
    override fun getPlanning(): Observable<List<PlanningEntity>> {
        return Observable.create { e ->
            val plannings: MutableList<PlanningEntity> = ArrayList()
            val rawData = PlanningDataHandler.instance.getPlannings()
            if (rawData.size() > 0) {
                for (i in 1..rawData.size()) {
                    plannings.add(PlanningEntity(i, rawData[i]))
                }
            }
            e.onNext(plannings)
            e.onComplete()
        }
    }

    override fun savePlanning(plannings: List<PlanningEntity>): Observable<String> {
        return Observable.create { e ->
            PlanningDataHandler.instance.savePlannings(plannings)
            e.onNext("")
            e.onComplete()
        }
    }

    override fun addPlanning(planning: PlanningEntity): Observable<String> {
        return Observable.create { e ->
            PlanningDataHandler.instance.addPlanning(planning)
            e.onNext("")
            e.onComplete()
        }
    }
}